File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/entity/ViewInterfaceInfo.java
package com.omegar.mvp.compiler.entity;

import com.omegar.mvp.MvpProcessor;
import com.omegar.mvp.compiler.Util;
import com.squareup.kotlinpoet.ClassName;
import com.squareup.kotlinpoet.ClassNames;
import com.squareup.kotlinpoet.ParameterizedTypeName;
import com.squareup.kotlinpoet.TypeName;
import com.squareup.kotlinpoet.TypeVariableName;
import com.squareup.kotlinpoet.TypeVariableNames;

import java.util.List;
import java.util.stream.Collectors;

import javax.annotation.Nullable;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;

/**
 * Date: 27-Jul-2017
 * Time: 13:04
 *
 * @author Evgeny Kursakov
 */
@SuppressWarnings("NewApi")
public class ViewInterfaceInfo implements TypeElementHolder {
    @Nullable
    private final TypeElement mSuperInterfaceType;
    private final TypeElement mElement;
    private final ClassName mName;
    private final List<TypeVariableName> mTypeVariables;
    private final List<TypeVariableName> mParentTypeVariables;

    private final List<ViewCommandInfo> mCommands;

    public ViewInterfaceInfo(@Nullable TypeElement superInterfaceType,
                             TypeElement element,
                             List<ViewCommandInfo> commands,
                             List<TypeVariableName> typeVariables,
                             List<TypeVariableName> parentTypeVariables
    ) {
        mSuperInterfaceType = superInterfaceType;
        mElement = element;
        mName = ClassNames.get(element);
        mCommands = commands;
        mTypeVariables = typeVariables;
        mParentTypeVariables = parentTypeVariables;
    }

    @Nullable
    public TypeElement getSuperInterfaceType() {
        return mSuperInterfaceType;
    }

    public TypeElement getTypeElement() {
        return mElement;
    }

    public ClassName getName() {
        return mName;
    }

    public TypeName getNameWithTypeVariables() {
        if (mTypeVariables.isEmpty()) {
            return mName;
        } else {
            TypeVariableName[] names = new TypeVariableName[mTypeVariables.size()];
            mTypeVariables.toArray(names);

            return ParameterizedTypeName.get(mName, names);
        }
    }

    public List<TypeVariableName> getTypeVariables() {
        return mTypeVariables;
    }

    public List<TypeVariableName> getParentTypeVariables() {
        return mParentTypeVariables;
    }

    public List<ViewCommandInfo> getCommands() {
        return mCommands;
    }

    public String getViewStateFullName(Elements elements) {
        return getViewStateFullName(elements, mElement);
    }

    public String getViewStateSimpleName(Elements elements) {
        return getViewStateSimpleName(elements, mElement);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ViewInterfaceInfo that = (ViewInterfaceInfo) o;

        return mName != null ? mName.equals(that.mName) : that.mName == null;
    }

    @Override
    public int hashCode() {
        return mName != null ? mName.hashCode() : 0;
    }

    @Override
    public String toString() {
        return "ViewInterfaceInfo{" +
                "superInterfaceInfo=" + mSuperInterfaceType +
                ", element=" + mElement +
                '}';
    }

    public static String getViewStateFullName(Elements elements, TypeElement viewTypeElement) {
        return Util.getFullClassName(elements, viewTypeElement) + MvpProcessor.VIEW_STATE_SUFFIX;
    }

    public static String getViewStateFullName(String fullViewName) {
        return fullViewName + MvpProcessor.VIEW_STATE_SUFFIX;
    }

    public static String getViewFullName(Elements elements, TypeElement viewTypeElement) {
        return Util.getFullClassName(elements, viewTypeElement);
    }

    public static String getViewStateSimpleName(Elements elements, TypeElement viewTypeElement) {
        return Util.getSimpleClassName(elements, viewTypeElement) + MvpProcessor.VIEW_STATE_SUFFIX;
    }


}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/entity/ViewInterfaceInfo.kt
package com.omegar.mvp.compiler.entity

import com.omegar.mvp.MvpProcessor
import com.omegar.mvp.compiler.MoxyConst
import com.omegar.mvp.compiler.Util
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.TypeVariableName
import com.squareup.kotlinpoet.asClassName
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements

/**
 * Date: 27-Jul-2017
 * Time: 13:04
 *
 * @author Evgeny Kursakov
 */
class ViewInterfaceInfo(
    val superInterfaceType: TypeElement?,
    override val typeElement: TypeElement,
    val commands: List<ViewCommandInfo>,
    val typeVariables: List<TypeVariableName>,
    val parentTypeVariables: List<TypeVariableName>
) : TypeElementHolder {

    companion object {
        fun getViewStateFullName(elements: Elements?, viewTypeElement: TypeElement?): String {
            return Util.getFullClassName(elements, viewTypeElement) + MoxyConst.VIEW_STATE_SUFFIX
        }

        @JvmStatic
        fun getViewStateFullName(fullViewName: String): String {
            return fullViewName + MoxyConst.VIEW_STATE_SUFFIX
        }

        @JvmStatic
        fun getViewFullName(elements: Elements?, viewTypeElement: TypeElement?): String {
            return Util.getFullClassName(elements, viewTypeElement)
        }

        fun getViewStateSimpleName(elements: Elements?, viewTypeElement: TypeElement?): String {
            return Util.getSimpleClassName(elements, viewTypeElement) + MoxyConst.VIEW_STATE_SUFFIX
        }
    }

    val name: ClassName = typeElement.asClassName()

    val nameWithTypeVariables: TypeName = if (typeVariables.isEmpty()) name else name.parameterizedBy(typeVariables)

    fun getViewStateFullName(elements: Elements?): String {
        return getViewStateFullName(elements, typeElement)
    }

    fun getViewStateSimpleName(elements: Elements?): String {
        return getViewStateSimpleName(elements, typeElement)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as ViewInterfaceInfo
        return name == that.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    override fun toString(): String {
        return "ViewInterfaceInfo{" +
                "superInterfaceInfo=" + superInterfaceType +
                ", element=" + typeElement +
                '}'
    }

}